enable Greenplum Clusters to run across subnets with mirrors

dkrieger-pvtl opened the pull request:
The three CM utilities gpinitsystem, gpaddmirrors, and gpmovemirrors
now have the relevant pg_hba.conf entries to allow WAL replication to
mirrors from their respective primaries. (gpinitstandby already had
this feature)

There are two parts to this commit:
1). modify the CM utilities to add the pg_hba.conf entries to
allow WAL replication to mirrors across a subnet.
2). test the relevant CM utilities across subnets

For testing, we created a cluster with two subnets in Concourse.  Then
we wrote Behave tests to validate the utilities. 

Note that the original implementation modified the cluster generation code so that
a bash line can distinguish between "mdw-1" and "mdw-2".  Instead, we modify the
test pipeline code(in this repo) to call the clusters "mdw-1" and "smdw-2".

NOTE: a full pipeline run is pending.

Co-authored-by: Jacob Champion <pchampion@pivotal.io>
Co-authored-by: Bhuvnesh Chaudhary <bchaudhary@pivotal.io>
Co-authored-by: Kalen Krempely <kkrempely@pivotal.io>
Co-authored-by: David Krieger <dkrieger@pivotal.io>
Co-authored-by: Jamie McAtamney <jmcatamney@pivotal.io>
Co-authored-by: Mark Sliva <msliva@pivotal.io>

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

jchampio commented on gpMgmt/bin/gpmovemirrors:
```diff
@@ -370,25 +370,32 @@ try:
+                        entry = "host all {0} {1} trust".format(unix.getUserName(), cidr)
+                        entries.append(entry)
+                        entry = 'host  replication {0} {1} trust'.format(unix.getUserName(), cidr)
+                        entries.append(entry)
```
This looks like it needs to be pulled out into a common path, after we've determined whether to use a single hostname or a list of CIDR blocks.

marksliva replied:
Can you explain what you mean by this? We're not sure what you mean exactly.

marksliva replied:
I pushed a change to this and `gpaddmirrors` and `gpexpand`.

jchampio commented on gpMgmt/bin/gpmovemirrors:
```diff
@@ -370,25 +370,32 @@ try:
+                        entry = 'host  replication {0} {1} trust'.format(unix.getUserName(), cidr)
+                        entries.append(entry)
+
+                cmdStr = ". {0}/greenplum_path.sh; echo '{1}' >> {2}/pg_hba.conf; pg_ctl -D {2} reload".format(
```
As a general note, not something for this PR -- switching to `subprocess` would allow us to stream the file contents through stdin instead of using this strange `echo` pattern.

marksliva replied:
Good point. We'll keep that in mind for future refactors.

jchampio commented on gpMgmt/bin/gpmovemirrors:
```diff
@@ -370,25 +370,32 @@ try:
+
+                cmdStr = ". {0}/greenplum_path.sh; echo '{1}' >> {2}/pg_hba.conf; pg_ctl -D {2} reload".format(
+                    os.environ["GPHOME"],
+                    "\n".join(entries),
```
This massive entry string needs to be escaped. `pipes.quote()` will do that and then you can remove the single quotes around the argument to `echo` above.

marksliva replied:
Done.

jchampio commented on gpMgmt/bin/gppylib/programs/clsAddMirrors.py:
```diff
@@ -501,23 +501,28 @@ def validate_heap_checksums(self, gpArray):
         logger.info("Starting to modify pg_hba.conf on primary segments to allow replication connections")
-        replicationStr = ". {0}/greenplum_path.sh; echo 'host  replication {1} samenet trust {2}' >> {3}/pg_hba.conf; pg_ctl -D {3} reload"
-
+        entries = []
```
Same notes as for `gpmovemirrors` above.

marksliva replied:
This needs a change. `entries` should be empty at the beginning of each iteration of the segmentPair loop.

marksliva replied:
I moved the entries declaration to the beginning of the loop.

jchampio commented on gpMgmt/bin/lib/gpcreateseg.sh:
```diff
@@ -229,7 +229,19 @@ CREATE_QES_MIRROR () {
+        local CIDR_ADDR
+        for ADDR in "${MIRROR_IPV4_LOCAL_ADDRESS_ALL[@]}" "${MIRROR_IPV6_LOCAL_ADDRESS_ALL[@]}"; do
+            CIDR_ADDR=$(GET_CIDRADDR $ADDR)
+            RUN_COMMAND_REMOTE ${PRIMARY_HOSTADDRESS} "${EXPORT_GPHOME}; . ${GPHOME}/greenplum_path.sh; echo 'host  replication ${GP_USER} ${CIDR_ADDR} trust' >> ${PRIMARY_DIR}/pg_hba.conf; pg_ctl -D ${PRIMARY_DIR} reload"
```
Running one `ssh` for every IP is one of the main contributors to segment creation slowness; we need to construct our additions once and then send them across.

marksliva replied:
Done. The character limit for echo is ARG_MAX (around 250K characters) and we don't anticipate that this will be hit.

bhuvnesh2703 commented on gpMgmt/test/behave/mgmt_utils/gpinitsystem.feature:
```diff
@@ -184,3 +184,8 @@ Feature: gpinitsystem tests
+#    The concourse tasks to bring up and verify the cross-subnet environment implciitly tests gpinitsystem.
+#    Therefore, no need to repeat the same test here.
+#    @cross_subnet
+#    Scenario: gpinitsystem succeeds across subnets
```
sure

bhuvnesh2703 commented on gpMgmt/test/behave/mgmt_utils/cross_subnet.feature:
```diff
@@ -0,0 +1,61 @@
+          And the standby replicates and fails over and back correctly
+
+    # Validate that gpmovemirrors can move mirrors across subnets
+      # TODO: check pg_hba.conf file for exact entries
```
removed the todo, as its unnecessary.. the test covers the required functionality.

bhuvnesh2703 commented on concourse/pipelines/templates/gpdb-tpl.yml:
```diff
@@ -1835,6 +1835,7 @@ jobs:
             vars:
               instance_type: n1-standard-2
               subnet: toolshed
+              host_name: 'smdw'
```
removed this line as we agreed to filter the first master hostname.

dkrieger-pvtl commented on gpMgmt/test/behave/mgmt_utils/steps/mgmt_utils.py:
```diff
@@ -2950,3 +2991,30 @@ def impl(context, command, input):
     context.stdout_message = stdout
     context.error_message = stderr
+
+# NOTE: when we update to Python 3, its ipaddress module would make this easier
```
I thought that when we switch to Py3, we'll require all hosts running gpdb to use Py3 for all GP commands.  Hence, I'm not sure what your comment is meant to convey.

jchampio replied:
I mean that `ip_address` doesn't fix the fact that we'd have to push a Python script across SSH to be able to use it, and when we tried to do that with Python 2 we decided it was easier to just run a Bash pipeline.

marksliva replied:
The `NOTE:` was deleted, since we are now doing this in bash. Although if we were using Python 3, the use of the library could result in Python being concise enough to be executed remotely.

dkrieger-pvtl commented on concourse/pipelines/templates/gpdb-tpl.yml:
```diff
@@ -333,6 +335,49 @@ resources:
       bucket: gpdb5-pipeline-dynamic-terraform
       bucket_path: clusters-google/
 
+# todo: make this an anchor and use that once anchors are defined before this point
```
We used an anchor for `terraform`, `terraform_1` and `terraform_2`.  We had to move all the anchors to the top, but that seems like a good idea anyway.

dkrieger-pvtl commented on gpMgmt/test/behave/mgmt_utils/cross_subnet.feature:
```diff
@@ -0,0 +1,61 @@
+      Examples:
+        | hba_hostnames | options             |
+        |  0            |  use_ip_addresses   |
+        |  1            |  '--hba-hostnames'  |
```
> 1. Why the single-quotes around `--hba-hostnames`?

The single ticks are removed.  I thought that non-alphanumeric characters had to be escaped, and I did so preemptively.  But they are not needed.

> 2. Why the `use_ip_addresses` pseudo-option instead of just blank (since that's what we substitute anyway)?

Because I couldn't get it to work otherwise.  In short, using Behave's `Scenario Outline`, I recall trying ` `, and a few other combinations and it always did the wrong thing.  We'll try again.

dkrieger-pvtl commented on gpMgmt/test/behave/mgmt_utils/environment.py:
```diff
@@ -19,7 +19,7 @@ def before_all(context):
     # we should be able to run gpexpand without having a cluster initialized
     tags_to_skip = ['gpexpand', 'gpaddmirrors', 'gpstate', 'gpmovemirrors',
-                    'gpconfig', 'gpssh-exkeys', 'gpstop', 'gpinitsystem']
+                    'gpconfig', 'gpssh-exkeys', 'gpstop', 'gpinitsystem', 'cross_subnet']
```
We added a story to track this.

dkrieger-pvtl commented on gpMgmt/test/behave/mgmt_utils/steps/cross_subnet.py:
```diff
@@ -0,0 +1,85 @@
+    # For the 'standby' case, we set PGHOST back to its original value instead
+    # of 'mdw-1'.  When the function impl() is called, PGHOST is initially unset
+    # by the test framework, and we want to respect that.
+    # TODO: set it to 'mdw-1' at the end since that's what we want?
```
left in by mistake; resolved.

dkrieger-pvtl commented on concourse/pipelines/templates/gpdb-tpl.yml:
```diff
@@ -757,18 +802,26 @@ anchors:
   get_params:
     action: destroy
 
+# todo: can we re-use some of the above?
```
Done.

dkrieger-pvtl commented on gpMgmt/test/behave/mgmt_utils/steps/mirrors_mgmt_utils.py:
```diff
@@ -202,6 +207,55 @@ def impl(context):
 
 
+# TODO: this function might be simpler if we hard-code the two lines to add.
+@given('a gpmovemirrors cross_subnet input file is created')
```
Done.

BaiShaoqi commented on gpMgmt/bin/lib/gpcreateseg.sh:
```diff
@@ -229,7 +229,23 @@ CREATE_QES_MIRROR () {
+    else
+        PG_HBA_ENTRY="host  replication ${GP_USER} localhost trust\nhost  replication ${GP_USER} ${GP_HOSTADDRESS} trust"
+        RUN_COMMAND_REMOTE ${PRIMARY_HOSTADDRESS} "${EXPORT_GPHOME}; . ${GPHOME}/greenplum_path.sh; echo -e '$PG_HBA_ENTRY' >> ${PRIMARY_DIR}/pg_hba.conf; pg_ctl -D ${PRIMARY_DIR} reload"
+    fi
```
Tried the code change you made in the `gpMgmt/bin/lib/gpcreateseg.sh`, it worked car start mirrors on my cluster. before your code change, it can not start mirrors on my cluster, show error `20191223:09:56:24:016638 gpcreateseg.sh:shaoqi-mdw:gpadmin-[INFO]:-Commencing remote /bin/ssh shaoqi-sdw2 export GPHOME=/usr/local/greenplum-db/.; . /usr/local/greenplum-db/./greenplum_path.sh; rm -rf /data/mirror2/gpseg4; /usr/local/greenplum-db/./bin/pg_basebackup --xlog-method=stream --slot='internal_wal_replication_slot' -R -c fast -E ./db_dumps -E ./gpperfmon/data -E ./gpperfmon/logs -D /data/mirror2/gpseg4 -h shaoqi-sdw -p 6001 --target-gp-dbid 12;
pg_basebackup: could not connect to server: FATAL:  no pg_hba.conf entry for replication connection from host "10.128.15.205", user "gpadmin", SSL off` https://pivotal.slack.com/archives/C4B0FBK0T/p1577095593042500

dkrieger-pvtl replied:
Great, so is your entire issue reported in that Slack thread resolved by this PR?  You can also, of course, cherry-pick our entire PR into your branch...

dkrieger-pvtl replied:
^^^ @BaiShaoqi

BaiShaoqi replied:
@dkrieger-pvtl , yes, entire issue reported in that Slack thread resolved by this PR(but only applied the code your made in  `gpMgmt/bin/lib/gpcreateseg.sh`), will cherry-pick entire PR and try test more.

BaiShaoqi replied:
# gpaddmirrors test
1. start a cluster with no mirrors, 
```
postgres=# select * from gp_segment_configuration;
 dbid | content | role | preferred_role | mode | status | port |  hostname   |   address   |        datadir
------+---------+------+----------------+------+--------+------+-------------+-------------+-----------------------
    1 |      -1 | p    | p              | n    | u      | 5432 | shaoqi-mdw  | shaoqi-mdw  | /data/master/gpseg-1
    2 |       0 | p    | p              | n    | u      | 6000 | shaoqi-sdw2 | shaoqi-sdw2 | /data/primary1/gpseg0
    5 |       3 | p    | p              | n    | u      | 6000 | shaoqi-sdw  | shaoqi-sdw  | /data/primary1/gpseg3
    3 |       1 | p    | p              | n    | u      | 6001 | shaoqi-sdw2 | shaoqi-sdw2 | /data/primary2/gpseg1
    6 |       4 | p    | p              | n    | u      | 6001 | shaoqi-sdw  | shaoqi-sdw  | /data/primary2/gpseg4
    4 |       2 | p    | p              | n    | u      | 6002 | shaoqi-sdw2 | shaoqi-sdw2 | /data/primary3/gpseg2
    7 |       5 | p    | p              | n    | u      | 6002 | shaoqi-sdw  | shaoqi-sdw  | /data/primary3/gpseg5
    8 |      -1 | m    | m              | s    | u      | 5433 | shaoqi-mdw  | shaoqi-mdw  | /data/standby
(8 rows)
```
2. gpaddmirrors  to add two different host shaoqi-sdw3, shaoqi-sdw4,  the command run successfully without any error
```
[gpadmin@shaoqi-mdw ~]$ cat gpconfigs/mirror_config_file
0|shaoqi-sdw4|7000|/data/mirror1
1|shaoqi-sdw4|7001|/data/mirror2
2|shaoqi-sdw4|7002|/data/mirror3
3|shaoqi-sdw3|7000|/data/mirror1
4|shaoqi-sdw3|7001|/data/mirror2
5|shaoqi-sdw3|7002|/data/mirror3
[gpadmin@shaoqi-mdw ~]$ gpaddmirrors -i gpconfigs/mirror_config_file
```
3. check the mirros have been added to shaoq-sdw3, and shaoqi-sdw4
```
postgres=# select * from gp_segment_configuration;
 dbid | content | role | preferred_role | mode | status | port |  hostname   |   address   |        datadir
------+---------+------+----------------+------+--------+------+-------------+-------------+-----------------------
    1 |      -1 | p    | p              | n    | u      | 5432 | shaoqi-mdw  | shaoqi-mdw  | /data/master/gpseg-1
    2 |       0 | p    | p              | n    | u      | 6000 | shaoqi-sdw2 | shaoqi-sdw2 | /data/primary1/gpseg0
    5 |       3 | p    | p              | n    | u      | 6000 | shaoqi-sdw  | shaoqi-sdw  | /data/primary1/gpseg3
    3 |       1 | p    | p              | n    | u      | 6001 | shaoqi-sdw2 | shaoqi-sdw2 | /data/primary2/gpseg1
    6 |       4 | p    | p              | n    | u      | 6001 | shaoqi-sdw  | shaoqi-sdw  | /data/primary2/gpseg4
    4 |       2 | p    | p              | n    | u      | 6002 | shaoqi-sdw2 | shaoqi-sdw2 | /data/primary3/gpseg2
    7 |       5 | p    | p              | n    | u      | 6002 | shaoqi-sdw  | shaoqi-sdw  | /data/primary3/gpseg5
    8 |      -1 | m    | m              | s    | u      | 5433 | shaoqi-mdw  | shaoqi-mdw  | /data/standby
    9 |       0 | m    | m              | n    | d      | 7000 | shaoqi-sdw4 | shaoqi-sdw4 | /data/mirror1
   10 |       1 | m    | m              | n    | d      | 7001 | shaoqi-sdw4 | shaoqi-sdw4 | /data/mirror2
   11 |       2 | m    | m              | n    | d      | 7002 | shaoqi-sdw4 | shaoqi-sdw4 | /data/mirror3
   12 |       3 | m    | m              | n    | d      | 7000 | shaoqi-sdw3 | shaoqi-sdw3 | /data/mirror1
   13 |       4 | m    | m              | n    | d      | 7001 | shaoqi-sdw3 | shaoqi-sdw3 | /data/mirror2
   14 |       5 | m    | m              | n    | d      | 7002 | shaoqi-sdw3 | shaoqi-sdw3 | /data/mirror3
(14 rows)
```
# gpmovemirrors test
1. start a cluster with mirrors (only used there host: shaoqi-mdw, shaoqi-sdw, shaoqi-sdw2)
```
postgres=# select * from gp_segment_configuration;
 dbid | content | role | preferred_role | mode | status | port |  hostname   |   address   |        datadir
------+---------+------+----------------+------+--------+------+-------------+-------------+-----------------------
    1 |      -1 | p    | p              | n    | u      | 5432 | shaoqi-mdw  | shaoqi-mdw  | /data/master/gpseg-1
    3 |       1 | p    | p              | s    | u      | 6001 | shaoqi-sdw2 | shaoqi-sdw2 | /data/primary2/gpseg1
    9 |       1 | m    | m              | s    | u      | 7001 | shaoqi-sdw  | shaoqi-sdw  | /data/mirror2/gpseg1
    4 |       2 | p    | p              | s    | u      | 6002 | shaoqi-sdw2 | shaoqi-sdw2 | /data/primary3/gpseg2
   10 |       2 | m    | m              | s    | u      | 7002 | shaoqi-sdw  | shaoqi-sdw  | /data/mirror3/gpseg2
    2 |       0 | p    | p              | s    | u      | 6000 | shaoqi-sdw2 | shaoqi-sdw2 | /data/primary1/gpseg0
    8 |       0 | m    | m              | s    | u      | 7000 | shaoqi-sdw  | shaoqi-sdw  | /data/mirror1/gpseg0
    5 |       3 | p    | p              | s    | u      | 6000 | shaoqi-sdw  | shaoqi-sdw  | /data/primary1/gpseg3
   11 |       3 | m    | m              | s    | u      | 7000 | shaoqi-sdw2 | shaoqi-sdw2 | /data/mirror1/gpseg3
    7 |       5 | p    | p              | s    | u      | 6002 | shaoqi-sdw  | shaoqi-sdw  | /data/primary3/gpseg5
   13 |       5 | m    | m              | s    | u      | 7002 | shaoqi-sdw2 | shaoqi-sdw2 | /data/mirror3/gpseg5
    6 |       4 | p    | p              | s    | u      | 6001 | shaoqi-sdw  | shaoqi-sdw  | /data/primary2/gpseg4
   12 |       4 | m    | m              | s    | u      | 7001 | shaoqi-sdw2 | shaoqi-sdw2 | /data/mirror2/gpseg4
   14 |      -1 | m    | m              | s    | u      | 5433 | shaoqi-mdw  | shaoqi-mdw  | /data/standby
(14 rows)
```
2. gpmovemirros from shaoqi-sdw to shaoqi-sdw3(this host has never been used during gpinitsystem)
```
[gpadmin@shaoqi-mdw ~]$ cat gpconfigs/move_config_file
shaoqi-sdw|7000|/data/mirror1/gpseg0 shaoqi-sdw3|7000|/data/mirror1/gpseg0
shaoqi-sdw|7001|/data/mirror2/gpseg1 shaoqi-sdw3|7001|/data/mirror2/gpseg1
shaoqi-sdw|7002|/data/mirror3/gpseg2 shaoqi-sdw3|7002|/data/mirror3/gpseg2
[gpadmin@shaoqi-mdw ~]$ gpmovemirrors -i gpconfigs/move_config_file
```
3. check the mirrors has gone from shaoqi-sdw to shaoqi-sdw3
```
postgres=# select * from gp_segment_configuration;
 dbid | content | role | preferred_role | mode | status | port |  hostname   |   address   |        datadir
------+---------+------+----------------+------+--------+------+-------------+-------------+-----------------------
    1 |      -1 | p    | p              | n    | u      | 5432 | shaoqi-mdw  | shaoqi-mdw  | /data/master/gpseg-1
    5 |       3 | p    | p              | s    | u      | 6000 | shaoqi-sdw  | shaoqi-sdw  | /data/primary1/gpseg3
   11 |       3 | m    | m              | s    | u      | 7000 | shaoqi-sdw2 | shaoqi-sdw2 | /data/mirror1/gpseg3
    7 |       5 | p    | p              | s    | u      | 6002 | shaoqi-sdw  | shaoqi-sdw  | /data/primary3/gpseg5
   13 |       5 | m    | m              | s    | u      | 7002 | shaoqi-sdw2 | shaoqi-sdw2 | /data/mirror3/gpseg5
    6 |       4 | p    | p              | s    | u      | 6001 | shaoqi-sdw  | shaoqi-sdw  | /data/primary2/gpseg4
   12 |       4 | m    | m              | s    | u      | 7001 | shaoqi-sdw2 | shaoqi-sdw2 | /data/mirror2/gpseg4
   14 |      -1 | m    | m              | s    | u      | 5433 | shaoqi-mdw  | shaoqi-mdw  | /data/standby
    4 |       2 | p    | p              | s    | u      | 6002 | shaoqi-sdw2 | shaoqi-sdw2 | /data/primary3/gpseg2
   10 |       2 | m    | m              | s    | u      | 7002 | shaoqi-sdw3 | shaoqi-sdw3 | /data/mirror3/gpseg2
    3 |       1 | p    | p              | s    | u      | 6001 | shaoqi-sdw2 | shaoqi-sdw2 | /data/primary2/gpseg1
    9 |       1 | m    | m              | s    | u      | 7001 | shaoqi-sdw3 | shaoqi-sdw3 | /data/mirror2/gpseg1
    2 |       0 | p    | p              | s    | u      | 6000 | shaoqi-sdw2 | shaoqi-sdw2 | /data/primary1/gpseg0
    8 |       0 | m    | m              | s    | u      | 7000 | shaoqi-sdw3 | shaoqi-sdw3 | /data/mirror1/gpseg0
(14 rows)
```
# gpinitsystem test
it has been tested during `gpmovemirrors test`
when doing `1. start a cluster with mirrors `, because or else it will fail with error can not find entry in pg_hba.conf, like in the slack thread.

# conclusion
All the scripts `gpMgmt/bin/` are taking effective when add mirror, or move mirror, or createseg.

jchampio replied:
Thanks! I'll mark this conversation resolved, but please let us know if you run into any other bugs.

pivotal-ning-yu commented on gpMgmt/bin/lib/gpcreateseg.sh:
```diff
@@ -229,7 +229,23 @@ CREATE_QES_MIRROR () {
-    RUN_COMMAND_REMOTE ${PRIMARY_HOSTADDRESS} "${EXPORT_GPHOME}; . ${GPHOME}/greenplum_path.sh; echo 'host  replication ${GP_USER} samenet trust' >> ${PRIMARY_DIR}/pg_hba.conf; pg_ctl -D ${PRIMARY_DIR} reload"
+    if [ $HBA_HOSTNAMES -eq 0 ]; then
+        MIRROR_IPV4_LOCAL_ADDRESS_ALL=(`$TRUSTED_SHELL $GP_HOSTADDRESS "$IPV4_ADDR_LIST_CMD | $GREP inet | $AWK '{print \\$2}' | $CUT -d'/' -f1"`)
+        MIRROR_IPV6_LOCAL_ADDRESS_ALL=(`$TRUSTED_SHELL $GP_HOSTADDRESS "$IPV6_ADDR_LIST_CMD | $GREP inet6 | $AWK '{print \\$2}' | $CUT -d'/' -f1"`)
```
Looks like `MIRROR_IPV{4,6}_LOCAL_ADDRESS_ALL` are not used outside the function, please declare them as `local`.

And I think there is no need to make them arrays, we could simplify the code to `$($TRUSTED_SHELL ...)`, and `for ADDR in $MIRROR_IPV4_LOCAL_ADDRESS_ALL ...` later, which are easier to read.

At least please use `$(cmd)` instead of `` `cmd` ``, the later is hard to read and maintain, it's especially bad to use both of them in the same function.

pivotal-ning-yu replied:
We could also simple `$CUT -d'/'` to `$CUT -d/`

marksliva replied:
Sure. I have made these changes.

pivotal-ning-yu commented on gpMgmt/bin/lib/gpcreateseg.sh:
```diff
@@ -229,7 +229,23 @@ CREATE_QES_MIRROR () {
+        RUN_COMMAND_REMOTE ${PRIMARY_HOSTADDRESS} "${EXPORT_GPHOME}; . ${GPHOME}/greenplum_path.sh; echo -e '$PG_HBA_ENTRIES' >> ${PRIMARY_DIR}/pg_hba.conf; pg_ctl -D ${PRIMARY_DIR} reload"
+    else
+        PG_HBA_ENTRY="host  replication ${GP_USER} localhost trust\nhost  replication ${GP_USER} ${GP_HOSTADDRESS} trust"
+        RUN_COMMAND_REMOTE ${PRIMARY_HOSTADDRESS} "${EXPORT_GPHOME}; . ${GPHOME}/greenplum_path.sh; echo -e '$PG_HBA_ENTRY' >> ${PRIMARY_DIR}/pg_hba.conf; pg_ctl -D ${PRIMARY_DIR} reload"
```
It's better to move the `RUN_COMMAND_REMOTE` part out of the `IF` block, we only need to decide the entry/entries in the if-else-fi.

marksliva replied:
Good point. I moved it below the fi.

pivotal-ning-yu commented on gpMgmt/bin/gppylib/programs/clsAddMirrors.py:
```diff
@@ -501,23 +501,32 @@ def validate_heap_checksums(self, gpArray):
                     mirror_hostname, _, _ = socket.gethostbyaddr(segmentPair.mirrorDB.getSegmentHostName())
-                    hba_line_entry = "\nhost all {0} {1} trust".format(unix.getUserName(), mirror_hostname)
-                    allow_pair_hba_line_entries.append(hba_line_entry)
+                    hba_line_entry = "host all {0} {1} trust".format(unix.getUserName(), mirror_hostname)
```
Personally I prefer named formating args:
```python
"host all {user} {host} trust".format(user=unix.getUserName(), host=mirror_hostname)
```

For simple cases there is no need to name them at all:
```python
"host all {} {} trust".format(unix.getUserName(), mirror_hostname)
```

marksliva replied:
I agree. However, since image `gpdb6-centos6-test` has Python 2.6.6, this shortened syntax `"host all {} ..."` will not work because the short format args feature only works for Python 2.7+.
See: https://docs.python.org/2.7/library/string.html#format-string-syntax

So that leaves us with the option of using the named syntax, which ends up being a bit more verbose but easier to read what is being written to the pg_hba.conf file.

pivotal-ning-yu commented on gpMgmt/bin/gppylib/programs/clsAddMirrors.py:
```diff
@@ -501,23 +501,32 @@ def validate_heap_checksums(self, gpArray):
-                cmdStr = replicationStr.format(os.environ["GPHOME"], unix.getUserName(), " ".join(allow_pair_hba_line_entries), segmentPair.primaryDB.datadir)
+                        hba_line_entry = "host all {0} {1} trust".format(unix.getUserName(), cidr)
+                        entries.append(hba_line_entry)
+                        entries.append('host  replication {0} {1} trust'.format(unix.getUserName(), cidr))
```
So we are appending two entries, one for `all` and one for `replication`, the first is added via a temp variable, the second is added directly, this makes me uncomfortable ;), could we stick on one way?  Use temp variables for both, or add both directly.

pivotal-ning-yu replied:
I see that we are calling `unix.getUserName()` many times, I'm not a python expert so I don't know how fast it is, but would it be better to create a temp variable for it?

marksliva replied:
Good point on the unnecessary `hba_line_entry` variable. I inlined it.

Also good idea of setting a username variable once. It doesn't make sense to call that function 6 times.

pivotal-ning-yu commented on gpMgmt/bin/gpmovemirrors:
```diff
@@ -370,25 +371,32 @@ try:
+                entries = []
+                primary_datadir = update_segment[0]
+                primary_hostname = update_segment[1]
+                newMirror_hostname = update_segment[2]
```
How about write it as:
```python
for primary_datadir, primary_hostname, newMirror_hostname in update_segments:
```

marksliva replied:
OK. Changed it.

marksliva commented on gpMgmt/bin/gppylib/programs/clsAddMirrors.py:
```diff
@@ -501,23 +501,37 @@ def validate_heap_checksums(self, gpArray):
+                                   .format(username=username, mirror_hostname=mirror_hostname))
+                    entries.append('host  replication {username} {mirror_hostname} trust'
+                                   .format(username=username, mirror_hostname=mirror_hostname))
+                    if mirror_hostname != "localhost":
```
@bhuvnesh2703 Do we need to add the same logic to `gpmovemirrors` and `gpexpand`?

marksliva replied:
This was removed and is resolved by keeping the samenet replication entry.

marksliva commented on gpMgmt/bin/gppylib/programs/clsAddMirrors.py:
```diff
@@ -501,23 +501,37 @@ def validate_heap_checksums(self, gpArray):
+                                       .format(username=username))
                 else:
                     mirror_ips = unix.InterfaceAddrs.remote('get mirror ips', segmentPair.mirrorDB.getSegmentHostName())
+                    if '127.0.0.1' not in mirror_ips:
```
Should we also add an entry for IPv6 loopback `::1/128` if we detect that any of the ips are IPv6 and when none of them are the loopback address? And similarly check that there is at least 1 IPv4 address in the list of ips and no IPv4 loopback before adding the IPv4 loopback `127.0.0.1/32`?

marksliva replied:
When I run the gpaddmirrors.feature test locally, I get this error:
```
      pivotals-MacBook-Pro-2.local (dbid 5): pg_basebackup: could not connect to server: FATAL:  no pg_hba.conf entry for replication connection from host "::1", user "pivotal"
      pivotals-MacBook-Pro-2.local (dbid 6): pg_basebackup: could not connect to server: FATAL:  no pg_hba.conf entry for replication connection from host "::1", user "pivotal"
      pivotals-MacBook-Pro-2.local (dbid 7): pg_basebackup: could not connect to server: FATAL:  no pg_hba.conf entry for replication connection from host "::1", user "pivotal"
```

jchampio replied:
(Turns out this issue is bigger. The `InterfaceAddrs` implementation only gives us IPv4 back, so we are accidentally excluding IPv6 from replication.)

marksliva replied:
Done. We created the `ifaddrs` utility and put it into `libexec` to list all of the IPv4 and IPv6 addresses and we add a replication entry for each interface address.

jchampio commented on gpMgmt/test/behave/mgmt_utils/steps/mirrors_mgmt_utils.py:
```diff
@@ -208,46 +208,23 @@ def impl(context):
 # TODO: this function might be simpler if we hard-code the two lines to add.
```
TODO can be removed.

marksliva replied:
Done.

jchampio commented on gpMgmt/test/behave/mgmt_utils/steps/mirrors_mgmt_utils.py:
```diff
@@ -208,46 +208,23 @@ def impl(context):
-                                   'datadir': seg.datadir
-                                   },
-                      actual_segs)
+    actual_segs = map(lambda seg: "%s|%s|%s" % (seg.hostname, seg.port, seg.datadir), actual_segs)
```
Ah. Now that the intention above is a little clearer, let's make this one clearer too:
```py
actual_segs = [
    "%s|%s|%s" % (seg.hostname, seg.port, seg.datadir)
    for seg in segs
    if seg.role == 'm' and seg.content in [2, 3]
]
```
`map` and `filter` are mostly useful if you already have a predefined function to map or filter on, not if you have to define your own lambdas. And when you find yourself using both it's almost certainly going to be nicer to read a list comprehension.

marksliva replied:
Done.

jchampio commented on gpMgmt/bin/gppylib/commands/gp.py:
```diff
@@ -817,21 +817,22 @@ def __init__(self, name, file, optName, optVal, optType='string', ctxt=LOCAL, re
+        for host in hosts:
+            hba_content += "\nhost all {username} {host} trust".format(username=username, host=host)
+            hba_content += "\nhost replication {username} {host} trust".format(username=username, host=host)
         # You might think you can substitute the primary and mirror addresses
```
nit: need a new paragraph here, before the comment.

marksliva replied:
Done.

jchampio commented on gpMgmt/bin/lib/gpcreateseg.sh:
```diff
@@ -230,8 +230,8 @@ CREATE_QES_MIRROR () {
-        local MIRROR_IPV4_LOCAL_ADDRESS_ALL=$($TRUSTED_SHELL $GP_HOSTADDRESS "$IPV4_ADDR_LIST_CMD | $GREP inet | $AWK '{print \\$2}' | $CUT -d/ -f1")
-        local MIRROR_IPV6_LOCAL_ADDRESS_ALL=$($TRUSTED_SHELL $GP_HOSTADDRESS "$IPV6_ADDR_LIST_CMD | $GREP inet6 | $AWK '{print \\$2}' | $CUT -d/ -f1")
+        local MIRROR_IPV4_LOCAL_ADDRESS_ALL=`$TRUSTED_SHELL $GP_HOSTADDRESS "$IPV4_ADDR_LIST_CMD | $GREP inet | $AWK '{print \\$2}' | $CUT -d/ -f1"`
+        local MIRROR_IPV6_LOCAL_ADDRESS_ALL=`$TRUSTED_SHELL $GP_HOSTADDRESS "$IPV6_ADDR_LIST_CMD | $GREP inet6 | $AWK '{print \\$2}' | $CUT -d/ -f1"`
```
This seems like a step backwards? Why use backticks?

marksliva replied:
We ended up leaving this code alone, so this comment is no longer applicable.

marksliva commented on gpMgmt/test/behave/mgmt_utils/steps/mirrors_mgmt_utils.py:
```diff
@@ -202,6 +205,33 @@ def impl(context):
             % (key, old_content_to_host[key], curr_content_to_host[key]))
 
 
+# TODO: this function might be simpler if we hard-code the two lines to add.
```
Delete this TODO.

marksliva replied:
Done.

jchampio commented on gpMgmt/bin/gppylib/programs/clsAddMirrors.py:
```diff
@@ -501,23 +501,35 @@ def validate_heap_checksums(self, gpArray):
+                for host in hosts:
+                    entries.append("host all {username} {host} trust".format(username=username, host=host))
+                    entries.append('host  replication {username} {host} trust'.format(username=username, host=host))
+                cmdStr = ". {gphome}/greenplum_path.sh; echo {entries} >> {primarydb_datadir}/pg_hba.conf; pg_ctl -D {primarydb_datadir} reload" \
```
nit: be careful to choose consise names when using named format fields, to avoid blowing up the line. E.g. `{datadir}` would be fine, and then you can look below to see that `segmentPair.primaryDB.datadir` is the particular data directory in use here.

marksliva replied:
Done.

jchampio commented on gpMgmt/bin/lib/gpcreateseg.sh:
```diff
@@ -229,7 +229,22 @@ CREATE_QES_MIRROR () {
+    else
+        PG_HBA_ENTRIES="host  replication ${GP_USER} localhost trust\nhost  replication ${GP_USER} ${GP_HOSTADDRESS} trust"
+    fi
+    RUN_COMMAND_REMOTE ${PRIMARY_HOSTADDRESS} "${EXPORT_GPHOME}; . ${GPHOME}/greenplum_path.sh; echo -e '$PG_HBA_ENTRIES' >> ${PRIMARY_DIR}/pg_hba.conf; pg_ctl -D ${PRIMARY_DIR} reload"
```
Shouldn't need `echo` here; we can stream the contents via stdin.

marksliva replied:
Were you thinking something like this?
```
RUN_COMMAND_REMOTE ${PRIMARY_HOSTADDRESS} "${EXPORT_GPHOME}; . ${GPHOME}/greenplum_path.sh; cat - >> ${PRIMARY_DIR}/pg_hba.conf; pg_ctl -D ${PRIMARY_DIR} reload" <<EOF
$PG_HBA_ENTRIES
EOF
```

What we're seeing is a redirection issue, because the function evaluates it to this:
```
ssh localhost "export GPHOME=/usr/local/gpdb; . "${GPHOME}"/greenplum_path.sh; cat - >> /datadir/pg_hba.conf; pg_ctl -D ${PRIMARY_DIR} reload" <<EOF
$PG_HBA_ENTRIES
EOF >> /tmp/logfile 2>&1
```
And it fails because of the `>> /tmp/logfile 2>&1`. We would rather not rewrite the `RUN_COMMAND_REMOTE` function. Is this what you had in mind?

jchampio replied:
Yeah. As one final point, whenever you write
```bash
cmd <<EOF
$var
EOF
```
remember that [herestrings](https://www.gnu.org/savannah-checkouts/gnu/bash/manual/bash.html#Here-Strings) are a thing:
```bash
cmd <<< "$var"
```

marksliva replied:
Good point. We changed it to a herestring.

paul-guo- commented on gpMgmt/bin/gppylib/commands/gp.py:
```diff
@@ -9,6 +9,7 @@
 import os, pickle, base64, time
 import os.path
 import pipes
+import subprocess
```
Please use subprocess32 in production
```
try:
    import subprocess32 as subprocess
except:
    import subprocess
```
Here is the context.
```commit 9c4a885b5523bbaad2e07b63ac6aa9007f5a270f
Author: Paul Guo <paulguo@gmail.com>
Date:   Fri Sep 20 11:25:31 2019 +0800

    Ship subprocess32 and replace subprocess with it in python code (#8658)

    * Ship modified python module subprocess32 again

    subprocess32 is preferred over subprocess according to python documentation.
    In addition we long ago modified the code to use vfork() against fork() to
    avoid some "Cannot allocate memory" kind of error (false alarm though - memory
    is actually sufficient) on gpdb product environment that is usually with memory
    overcommit disabled. 
```

marksliva replied:
Done.

bhuvnesh2703 commented on gpMgmt/bin/gpinitsystem:
```diff
@@ -465,11 +465,13 @@ CHK_PARAMS () {
 		ERROR_CHK $? "obtain IP address of Master host" 2
 		MASTER_IPV6_LOCAL_ADDRESS_ALL=(`$IPV6_ADDR_LIST_CMD | $GREP inet6 | $AWK '{print $2}' | $CUT -d'/' -f1`)
 		MASTER_IP_ADDRESS=(`$ECHO ${MASTER_IP_ADDRESS_ALL[@]} ${MASTER_IPV6_LOCAL_ADDRESS_ALL[@]}|$TR ' ' '\n'|$SORT -u|$TR '\n' ' '`)
+		MASTER_IP_ADDRESS_NO_LOOPBACK=($("$GPHOME"/libexec/ifaddrs --no-loopback))
```
where is this used?

marksliva replied:
It's not a good convention, but because we wanted to group the code next to the existing code, we defined it here, and it gets used in `gp_bash_functions.sh`.
Search for `MASTER_IPV6_LOCAL_ADDRESS_ALL` and you'll see the convention we're following.

marksliva replied:
We can move this if you have a strong opinion on moving it.

marksliva replied:
We can move this into `gp_bash_functions.sh` if you think that makes more sense.

bhuvnesh2703 replied:
ok.. thats fine..

jchampio replied:
(My vote would be to move it where it's used. No need to stick to the original poor practices; it'll just make maintenance harder later.)

marksliva replied:
Done. Moved it to where it gets used.

dkrieger-pvtl commented:
@jchampio See the fixups I added...essentially, I changed from `samenet` to `samehost` to allow same-host mirrors(testing seems like the only reason to do this; production would have the primary and mirror on different hosts).   

For `gpinitstandby`, unlike the other 4 commands we modified, we do not add the primary's hostname to the pg_hba.conf.  Why?  As I understand it, we add the primary's hostname to pg_hba.conf before creating a replica so that when that replica is promoted, _it_ can create a replacement replica that allows WAL replication.

jchampio commented on gpMgmt/bin/ifaddrs/main.c:
```diff
@@ -0,0 +1,105 @@
+#include <stdio.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <sys/types.h>
```
Ah, one final thing that needs to be done here to is to make sure that GPDB's `autoconf` setup reflects any new dependencies we've added in this list.

marksliva replied:
Ok. We updated configure.in to now check for inet_htop and getifaddrs.

jchampio commented on gpMgmt/bin/lib/gpcreateseg.sh:
```diff
@@ -229,7 +229,22 @@ CREATE_QES_MIRROR () {
+        do
+            CIDR_ADDR=$(GET_CIDRADDR $ADDR)
+            PG_HBA_ENTRIES="${PG_HBA_ENTRIES}"$'\n'"host  replication ${GP_USER} ${CIDR_ADDR} trust"
+        done
```
1) Is the deduplication necessary? It seems to add a lot of complexity.
2) The order of operations seems to be
- call `ifaddrs` which prints one interface per line
- convert that output into an array, losing the newlines
- convert that array into a space-separated string
- replace the spaces with newlines with `tr`
- deduplicate with `sort`
- replace the newlines with spaces with `tr`
- convert the string back to an array, losing the spaces
- convert the array back to a space-separated string with `${[@]}`

Let's streamline that a little bit. 😄 The array juggling is probably not helping very much.

marksliva replied:
Good point. We removed the deduplication.

jchampio commented on gpMgmt/bin/gpmovemirrors:
```diff
@@ -370,25 +370,35 @@ try:
+                    entries.append("host all {username} {hostname} trust".format(username=unix.getUserName(), hostname=hostname))
+                    entries.append("host replication {username} {hostname} trust".format(username=unix.getUserName(), hostname=hostname))
                 else:
                     newMirror_ips = unix.InterfaceAddrs.remote('get mirror ips', newMirror_hostname)
```
Is there a follow-up chore somewhere to replace `InterfaceAddrs` for the other cases?

dkrieger-pvtl replied:
There's a TODO in unix.py::InterfaceAddrs(): `# todo: This class should be replaced with gp.IfAddrs`.  We also added a story to track it.

jchampio commented on gpMgmt/bin/gpinitstandby:
```diff
@@ -509,31 +509,40 @@ def update_pg_hba_conf(options, array):
+        # Add samenet replication to support single-host development.
+        add_entry_to_new_section(['host', 'replication', current_user, 'samenet', 'trust'])
+        if not options.hba_hostnames:
+            if_addrs = gp.IfAddrs.list_addrs(options.standby_host)
```
Where is `gp.IfAddrs` defined?

dkrieger-pvtl replied:
See https://github.com/greenplum-db/gpdb/pull/9272#discussion_r368032258

jchampio commented on gpMgmt/bin/gppylib/commands/gp.py:
```diff
@@ -1565,6 +1577,21 @@ def __init__(self, name, options = "", ctxt = LOCAL, remoteHost = None):
        cmdStr = "$GPHOME/bin/gprecoverseg %s" % (options)
        Command.__init__(self,name,cmdStr,ctxt,remoteHost)
 
+class IfAddrs:
```
Oh, it's here. This needs to be moved to the first commit that uses it.

dkrieger-pvtl replied:
Will do.  Our technique is to commit two fixups: one that removes it from the gpexpand commit and the other adds it to the main "fix replication" commit.  Then, during rebase, we delete the gpexpand fixup. 

Perhaps there's a cleaner technique, but this works.

jchampio commented on gpMgmt/bin/gppylib/commands/gp.py:
```diff
@@ -1565,6 +1577,21 @@ def __init__(self, name, options = "", ctxt = LOCAL, remoteHost = None):
 
+class IfAddrs:
+    @staticmethod
+    def list_addrs(hostname=None, include_loopback=False):
```
Need some paragraphs in this method.

dkrieger-pvtl replied:
Done.

jchampio commented on gpMgmt/bin/gppylib/commands/gp.py:
```diff
@@ -1565,6 +1577,21 @@ def __init__(self, name, options = "", ctxt = LOCAL, remoteHost = None):
+            args.append('--no-loopback')
+        process = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        out, err = process.communicate()
+        if err:
```
`process.returncode` is what you need to check. A subprocess might terminate without stderr, or vice-versa.

marksliva replied:
We now use subprocess.check_output, and the error handling that interface provides.

jchampio commented on gpMgmt/bin/gppylib/commands/gp.py:
```diff
@@ -1565,6 +1577,21 @@ def __init__(self, name, options = "", ctxt = LOCAL, remoteHost = None):
+        process = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        out, err = process.communicate()
+        if err:
+            raise Exception("Command: `{args}` Error: {err}".format(args=' '.join(args), err=str(err)))
```
Is there a reason not to use `subprocess.check_output()` instead of rolling this ourselves?

marksliva replied:
Done.

jchampio replied:
What does our new `Exception` give us that the `check_output()` exception does not?

jchampio replied:
We'll remove the wrapper.

jchampio commented on gpMgmt/bin/gppylib/commands/gp.py:
```diff
@@ -1565,6 +1577,21 @@ def __init__(self, name, options = "", ctxt = LOCAL, remoteHost = None):
+        args = []
+        if hostname:
+            args.extend(['ssh', hostname])
+        ifaddrs = '%s/libexec/ifaddrs' % GPHOME
```
Can we trust the assignment of this `GPHOME` variable?

dkrieger-pvtl replied:
Do you mean `GPHOME=os.environ.get('GPHOME')` set as a global at module load time might actually change by the time we run this function?  I thought that a customer is not supposed to modify env variables after starting the database, nor are we.

jchampio replied:
As long as we've verified that we don't actually modify that `GPHOME` variable in terrible ways, I'm happy.

dkrieger-pvtl commented:
> @jchampio See the fixups I added...essentially, I changed from `samenet` to `samehost` to allow same-host mirrors(testing seems like the only reason to do this; production would have the primary and mirror on different hosts).
> 
> For `gpinitstandby`, unlike the other 4 commands we modified, we do not add the primary's hostname to the pg_hba.conf. Why? As I understand it, we add the primary's hostname to pg_hba.conf before creating a replica so that when that replica is promoted, _it_ can create a replacement replica that allows WAL replication.

@jchampio my question refers to this table: 
![utillity_replication_additions](https://user-images.githubusercontent.com/42985115/72643223-0e03f380-3923-11ea-90be-6221dc290ca3.png)

For gpinitstandby, houldn't we add a replication entry for the master into the standby?  It seems so to me, so when we failover and promote the standby, run gpinitstandby to recreate a standby on the original master host, and then gprecoverseg rebalance, we need that "reverse mapping".

dkrieger-pvtl commented:
We discussed https://github.com/greenplum-db/gpdb/pull/9272#issuecomment-575783954.  It turns out that `gpinitstandby` does NOT need to have the primary explicitly added to itself because gpinitsystem adds these mappings.  So, gpinitsystem has the mapping, and then gpinitstandby ends up copying over its pg_hba.conf via pg_basebackup.  So the standby ends up always having the primary mapping...

jchampio commented on gpMgmt/bin/lib/gp_bash_functions.sh:
```diff
@@ -723,8 +723,27 @@ BUILD_MASTER_PG_HBA_FILE () {
+        $ECHO "host     replication $USER_NAME         samehost       trust" >> ${GP_DIR}/$PG_HBA
+        if [ $HBA_HOSTNAMES -eq 0 ];then
+            local MASTER_IP_ADDRESS_NO_LOOPBACK=($("$GPHOME"/libexec/ifaddrs --no-loopback))
+            if [ x"" != x"$STANDBY_HOSTNAME" ] && [ "$STANDBY_HOSTNAME" != "$MASTER_HOSTNAME" ];then
```
nit: I don't think we support any platforms that still require this `x""`-style comparison.

jchampio commented on gpMgmt/test/behave/mgmt_utils/steps/tablespace_mgmt_utils.py:
```diff
@@ -203,3 +203,9 @@ def impl(context):
+def impl(context):
+    for tablespace in context.tablespaces.values():
+        tablespace.cleanup()
+    context.tablespaces = {}
```
Why is this here? Does it need to be moved to a post-scenario cleanup?

kalensk commented on concourse/pipelines/templates/gpdb-tpl.yml:
```diff
@@ -35,6 +35,143 @@
+    AWS_SECRET_ACCESS_KEY: {{tf-machine-secret-access-key}}
+    AWS_DEFAULT_REGION: {{tf-machine-region}}
+    BUCKET_PATH: clusters-google/
+    BUCKET_NAME: {{tf-bucket-name}}
```
super minor nit: I would move this down one, so it is above `set_failed_two_clusters ` which is where it is use.

kalensk commented on gpMgmt/bin/ifaddrs/main.c:
```diff
@@ -0,0 +1,105 @@
+				fprintf(stderr, "usage: %s [--no-loopback]\n", argv[0]);
+				return 1;
+		}
+	}
```
nit: Do we want to support `--help`?

jchampio replied:
I'm thinking not, for the reason you said.

kalensk commented on gpMgmt/bin/gppylib/commands/gp.py:
```diff
@@ -817,17 +821,25 @@ def __init__(self, name, file, optName, optVal, optType='string', ctxt=LOCAL, re
                     cidr_suffix = '/128' if ':' in ip else '/32'
                     cidr = ip + cidr_suffix
-                    hba_content += "\nhost all {0} {1} trust".format(getUserName(), cidr)
+                    hba_content += "\nhost all {username} {cidr} trust".format(username=username, cidr=cidr)
```
Can we also call `IfAddrs.list_addrs()` here instead of `InterfaceAddrs.remote()`?

```py
ips = InterfaceAddrs.remote('get mirror ips', address)
                for ip in ips:
                    cidr_suffix = '/128' if ':' in ip else '/32'
                    cidr = ip + cidr_suffix
                    hba_content += "\nhost all {username} {cidr} trust".format(username=username, cidr=cidr)
```

jchampio replied:
We'd like to eventually use `IfAddrs` everywhere. Backporting that change to a stable branch is a harder sell.

kalensk commented on gpMgmt/bin/gpinitstandby:
```diff
@@ -509,31 +509,40 @@ def update_pg_hba_conf(options, array):
-            if newline not in pg_hba_conf:
-                new_section.append(newline)
+        if not options.hba_hostnames:
+            for ip in standby_ips:
```
`standby_ips` is populated through `unix.InterfaceAddrs.remote`, should this be using `IfAddrs`? Why not?

jchampio replied:
See above.

kalensk commented on gpMgmt/bin/lib/gp_bash_functions.sh:
```diff
@@ -723,8 +723,27 @@ BUILD_MASTER_PG_HBA_FILE () {
 
         # Add replication config
+
+        # todo: do we still need this entry?
```
Can we address this comment or remove it?

jchampio replied:
Will remove.

kalensk commented on gpMgmt/test/behave/mgmt_utils/steps/cross_subnet.py:
```diff
@@ -0,0 +1,141 @@
+# 2). stop master/primary
+# 3). wait for automatic failover(for downed primary) or explicitly promote standby(for downed master)
+# 4). make sure data is on new master/primary
+# -). prepare for 6:
```
Uhhhh, why isn't this a `5).`? Seems confusing.

jchampio replied:
Will fix.

kalensk commented on gpMgmt/test/behave/mgmt_utils/cross_subnet.feature:
```diff
@@ -0,0 +1,62 @@
+Feature: Tests for a cross_subnet cluster
+
+    # Check that the hosts created in the cli_cross_subnet job are on different subnets.
+    Scenario: gpinitsystem works across subnets
```
Can these comments be rolled into the scenario description?

jchampio replied:
We'll remove them; in most cases they are incorrect.
